### DISCUSSION TITLE
Object Initialization Property Extension to Returned Objects

### DISCUSSION THREAD

**Discussion by hawkerm** [dotnet/csharplang#7883] [Language Ideas]
This would be a re-use of the [object initialization](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/object-and-collection-initializers#object-initializers) syntax beyond constructors for returned objects from methods. Would be similar to `with` with record types as well, I suppose.

Not sure how best to describe this, but it's something I always want to do when I get back objects from factories or other built-in API surfaces which return settings style objects that I'm trying to modify.

e.g. here's a scenario from the Windows API as an example customizing a bunch of paragraph settings from the [ITextParagraphFormat](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.text.itextparagraphformat) interface:

```cs
    var paragraphFormat = TextDocument.GetDefaultParagraphFormat();
    paragraphFormat.SpaceBefore = 24;
    paragraphFormat.SpaceAfter = 24;

    TextDocument.SetDefaultParagraphFormat(paragraphFormat);
```

What I'd prefer to do instead like if I had constructed this object is the following:

```cs
    var paragraphFormat = TextDocument.GetDefaultParagraphFormat() {
        SpaceBefore = 24,
        SpaceAfter = 24,
    };

    TextDocument.SetDefaultParagraphFormat(paragraphFormat);
```

This saves me from having to retype the variable name, or copy-paste a whole wall of text for increasing number of properties. Instead, the value of the object is more easily seen and grouped in a familiar construct.

```cs
    var newThing = SomeFactory.GetNewBlankFoo() {
       ConfigValue = 42,
       Name = "Bar",
       Timing = new(4.5),
       Alignment = Alignment.Right,
       WhatElse = "more...",
    };

    newThing.Execute();
```

This works well with the factory pattern as well, as sometimes that's the only way to get a 'new' construction of an object with some API surfaces, as a developer though I frequently want to modify those outcomes before use. Being able to use a familiar pattern I use for construction of other types would be a great feeling and shortcut to my productivity. Thanks!

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/blob/main/meetings/2020/LDM-2020-04-20.md

**Reply by hawkerm** [reply]
Thanks for the reference to a relevant discussion. I can see the concern with it only being for `new` objects. But I'd be equally happy with `with` and extending it beyond records as well, e.g.

```cs
    var paragraphFormat = TextDocument.GetDefaultParagraphFormat() with {
        SpaceBefore = 24,
        SpaceAfter = 24,
    };
```

Though, I think the question is would `with` be returning a second copy of the original return value or a modified instance of it. As it'd be more efficient in these cases to modify the original. With records, it's always a new 'instance' effectively, right?
